### PR TITLE
Add `from_pytorch` function to `FusionDefinition` API and remove number of dims based `define_tensor`

### PR DESCRIPTION
**Pull request by kevinstephano**
1. Adds a python based `FusionDefinition` class in order to enable defining `from_pytorch` from python.  This means the python bound C++ class is renamed to `_FusionDefinition` and is inherited as a parent class.
2. Removed number of dimensions based `define_tensor` function from `FusionDefinition`.  The number of dimensions based version was causing confusion as it only allows the user to specify non-broadcasted tensors.
3. Added the `from_pytorch` function to the `FusionDefinition`.
4. In order to enable `from_pytorch` added a `pytorch_utils.py` module to define the mapping between Pytorch date types and nvFuser data types.
5. Added a test to make sure `from_pytorch` is not allowed for CPU tensors.

Handles issue #2359.

### CODE REVIEW

File: third_party/nvfuser/python/pytorch_utils.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import torch
+from typing import Type, Union
+
+from ._C import DataType
+
+NumberTypeType = Union[Type[bool], Type[int], Type[float], Type[complex]]
+    
+_torch_dtype_to_nvfuser_dtype_map = {
```
Comment (jjsjann123) [inline third_party/nvfuser/python/pytorch_utils.py:8]: I'm a little bit uncomfortable with having another map defined here.

For python specific types, sure, but for torch types, maybe we want to expose `aten_to_data_type` under `third_party/nvfuser/csrc/type.h`

File: third_party/nvfuser/python/pytorch_utils.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    float: DataType.Double,
+    int: DataType.Int,
+    bool: DataType.Bool,
+}
+
+def torch_dtype_to_nvfuser_dtype(dtype: Union[torch.dtype, NumberTypeType]):
+    """
+    Translates from torch.dtype to nvFuser's DataType enum
+    """
+    return _torch_dtype_to_nvfuser_dtype_map[dtype]
```
Comment (jjsjann123) [inline third_party/nvfuser/python/pytorch_utils.py:30]: Nitpick: We should also return DataType::Null when it's an unsupported type. (other than a None). Which feels a little bit more consistent with cpp API :wink:

File: third_party/nvfuser/python/__init__.py [hunk 3]
```diff
@@ -1 +1,17 @@
...
+        """
+        try:
+            from .pytorch_utils import torch_dtype_to_nvfuser_dtype
+        except ImportError:
+            raise ImportError("Unable to import pytorch_utils!")
+
+        if not tensor.is_cuda:
+            raise ValueError("Tensor should be on a cuda device!")
+
+        return self.define_tensor(sizes=tensor.size(), strides=tensor.stride(),
```
Comment (jjsjann123) [inline third_party/nvfuser/python/__init__.py:34]: I think this is fine for now. I'm not a big fan of silently converting it to static size. I'm pretty sure user will just follow our example and come back complaining about static shapes later.

**Review by jjsjann123 — APPROVED** [review]
LGTM. Some minor comment

File: third_party/nvfuser/python/__init__.py (hunk 3 above)
Comment (kevinstephano) [reply third_party/nvfuser/python/__init__.py:34]: This API does not create static sizes.  It converts to symbolic sizes and determines contiguity underneath.

Comment (jjsjann123) [reply third_party/nvfuser/python/__init__.py:34]: ah, sorry I think we had this conversation before.. I keep forgetting it..

**Pull request merged**